src/annotation/decorator_utils.ts

```diff
@@ -1,68 +1,110 @@
 import * as ERROR_MSGS from "../constants/error_msgs";
 import * as METADATA_KEY from "../constants/metadata_keys";
 import { interfaces } from "../interfaces/interfaces";
+import { getFirstArrayDuplicate } from "../utils/js";
+
+function targetIsConstructorFunction<T = Object>(target:DecoratorTarget<T>): target is ConstructorFunction<T>{
+    return (target as any).prototype !== undefined;
+}
+
+interface ConstructorFunction<T = Object>{
+    new (...args:unknown[]): T,
+    prototype:T
```

Comment:
I think prototype should not be `T` since T is not an instance of the type, but the prototype instead

Reply:
![image](https://user-images.githubusercontent.com/11292998/118105412-43640880-b3d4-11eb-9a3f-1bcc914adaee.png)

Reply:
```
var ClassWithDecorators = /** @class */ (function () {
    function ClassWithDecorators(param) {
    }
    __decorate([
        decorator,
        __metadata("design:type", Dependency)
    ], ClassWithDecorators.prototype, "prop", void 0);
    __decorate([
        decorator,
        __metadata("design:type", Dependency)
    ], ClassWithDecorators, "staticProp", void 0);
    ClassWithDecorators = __decorate([
        __param(0, decorator),
        __metadata("design:paramtypes", [Dependency])
    ], ClassWithDecorators);
    return ClassWithDecorators;
}());
```

Reply:
Mapped T to `Prototype<T>`, properties now optional.

```
type Prototype<T> = {
    [Property in keyof T ]:
        T[Property] extends Function?
            T[Property] :
            T[Property] | undefined
} & {constructor:Function}

interface ConstructorFunction<T = Object>{
    new (...args:unknown[]): T,
    prototype:Prototype<T>
}

export type DecoratorTarget<T = Object> = ConstructorFunction<T> |  Prototype<T>
```